### PR TITLE
Add validation check before executing editdonation and adddonation

### LINKED ISSUE #105: Add validation check before adding new donation record OR editing existing donation record

### PR DESCRIPTION
**Pull request by adityagk753**
Fixes #105 

## Changes:
1. Modify constructor for `IsEligibleToDonatePredicate` to accept a `donationDate` parameter, for more explicit behaviour (instead of assuming that it always uses the current date).
2. Shift testing logic from `IsEligibleToDonatePredicate.java` to `validate()` method in `DonationRecord.java`. `IsEligibleToDonatePredicate`'s `.test()` method now makes use of `DonationRecord`'s `.validate()` method.
3. Add check for predecessor and successor donation records being ≥ 84 days away from a donation record that is being added or a donation record that is being edited.
4. Improve readability of test logic by removal of highly nested if-else conditions.
5. Fix test cases to be compatible with updated behaviour.  
    * Notably, many test cases were using a `BloodNet` that only contained a single donation record (but no person associated to that donation record).  
    * These tests failed because now, we are creating a `DonationRecord` object with the `personId`, and upon calling `.validate()` method of the donationRecord, we try to find the person object associated with that `personId` in the `BloodNet`, to extract the person's `dateOfBirth` (this is needed for validation).  
    * I have modified these test cases such that if there is a donation record in BloodNet, there should also be a Person object corresponding to that donation record in the BloodNet.

---

## Some useful tests to test edge cases (with `addcommand`):

- First, delete your `data/` folder so that the data is reset to default data with 6 people.
- Execute the below tests in order. 

---

* **Setup to test condition 1:**  
  - `add n/Guy1 p/88888888 b/B+ d/10-02-1995 e/guy@gmail.com`

* **Condition 1:**  
  - `adddonation p/7 d/09-02-2011 v/200` (Fail as person is only 15 years and 364 days old)  
  - `adddonation p/7 d/10-02-2011 v/200` (Pass as person just turned 16 on this date)  

---

* **Setup to test condition 2:**  
  - `add n/Guy2 p/88888888 b/B+ d/10-02-1995 e/guy@gmail.com`  
  - `adddonation p/8 d/11-02-2011 v/200` (Pass as person just turned 16 years and 1 days old on this date)  

* **Condition 2:**  
  - `adddonation p/8 d/05-05-2011 v/200` (Fail as exactly 83 days after predecessor)  
  - `adddonation p/8 d/06-05-2011 v/200` (Pass as exactly 84 days after predecessor)  

---

* **Setup to test condition 3:**  
  - `add n/Guy3 p/88888888 b/B+ d/10-02-1995 e/guy@gmail.com`  
  - `adddonation p/9 d/11-10-2025 v/200` (Pass as person just turned 16 years and 1 days old on this date)  

* **Condition 3:**  
  - `adddonation p/9 d/20-07-2025 v/200` (Fail as exactly 83 days before successor)  
  - `adddonation p/9 d/19-07-2025 v/200` (Pass as exactly 84 days before successor)  

---

* **Setup to test condition 4:**  
  - `add n/Guy4 p/88888888 b/B+ d/10-02-1950 e/guy@gmail.com`  

* **Condition 4:**  
  - `adddonation p/10 d/10-02-2011 v/200` (Fail as person is exactly 61 years and 0 days old)  
  - `adddonation p/10 d/09-02-2011 v/200` (Pass as person is 60 years and 364 days old)  

---

* **Setup to test condition 5 Test1:**  
  - `add n/Guy5 p/88888888 b/B+ d/10-02-1950 e/guy@gmail.com`  
  - `adddonation p/11 d/10-02-1966 v/200` (Pass as person was exactly 16 years old)  

* **Condition 5 Test1:**  
  - `adddonation p/11 d/10-02-2016 v/200` (Fail as person is exactly 66 years old, not a first-time donor, but has no donation record in past 3 years)  

---

* **Setup to test condition 5 Test2:**  
  - `add n/Guy6 p/88888888 b/B+ d/10-02-1950 e/guy@gmail.com`  
  - `adddonation p/12 d/10-02-2010 v/200` (Pass as the person just turned 60 years old which is fine)  

* **Condition 5 Test2:**  
  - `adddonation p/12 d/10-02-2013 v/200` (Pass as the person just turned 63 years old, but he donated in past 3 years (must have donated on a date >= 10-02-2010), on 10-02-2010)  
  - `adddonation p/12 d/11-02-2016 v/200` (Fail as the person just turned 66 years old and 0 days, but he did not donate in past 3 years (did not donate on a date >= 12-02-2013))  
  - `adddonation p/12 d/09-02-2016 v/200` (Pass as the person is 65 years old and 364 days, and he donated in past 3 years (must have donated on a date >= 10-02-2013), on 10-02-2013)  
  - `adddonation p/12 d/09-02-2019 v/200` (Fail as person is non-first-time donor and age is more than 66 years old and he did not donate in the past 3 years (must have donated on a date >= 10-02-2016, but his most recent one is 09-02-2016))  
  - `adddonation p/12 d/08-02-2019 v/200` (Pass as person is non-first-time donor and age is more than 66 years old, but he donated in the past 3 years (must have donated on a date >= 09-02-2016), on 09-02-2016)

### CODE REVIEW

File: src/main/java/bloodnet/model/person/IsEligibleToDonatePredicate.java [hunk 1]
```diff
@@ -13,52 +16,119 @@
...
 
     private final Model model;
+    private final DonationDate donationDate;
     /**
      * Constructs a {@code IsEligibleToDonatePredicate}.
+     * Sets {@code donationDate} to today's date.
      */
     public IsEligibleToDonatePredicate(Model model) {
         this.model = model;
+
```
Comment (Eugene-Lek) [inline src/main/java/bloodnet/model/person/IsEligibleToDonatePredicate.java]: To avoid unexpected behaviour, it is better to require users of this predicate to explicitly specify the donation date.  In other words, we should remove the original constructor (the one which doesn't have the donation date as a param) and update Li Min's use of this predicate with the current date

**Review by Eugene-Lek — CHANGES_REQUESTED** [review]

File: src/main/java/bloodnet/logic/commands/EditDonationCommand.java [hunk 2]
```diff
@@ -40,6 +41,9 @@ public class EditDonationCommand extends Command {
     public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
     public static final String MESSAGE_DUPLICATE_DONATION_RECORD =
             "No change to the donation record.";
+    public static final String MESSAGE_VIOLATES_ELIBILITY_CRITERIA =
+            "This edit violates the donation eligibility criteria. "
```
[the remark this answers is not in the document]
Comment (adityagk753) [reply src/main/java/bloodnet/logic/commands/EditDonationCommand.java]: Nice idea
